rusk-wallet/src/clients.rs

```diff
@@ -248,133 +130,126 @@ impl StateStore {
     }
 
     pub async fn sync(&self) -> Result<(), Error> {
-        let store = self.store.clone();
+        sync_db(&self.client, &self.cache(), &self.addresses, self.status).await
+    }
+
+    /// Requests that a node prove the given transaction and later propagates it
+    pub fn propagate(&self, utx: Transaction) -> Result<Transaction, Error> {
         let status = self.status;
-        let (cache, client) = {
-            let state = self.inner.lock().unwrap();
+        let tx_bytes = utx.to_var_bytes();
 
-            let cache = state.cache.clone();
-            let client = state.client.clone();
-            (cache, client)
-        };
+        status("Attempt to preverify tx...");
+        let preverify_req = RuskRequest::new("preverify", tx_bytes.clone());
+        let _ = self.client.call(2, "rusk", &preverify_req).wait()?;
+        status("Preverify success!");
 
-        sync_db(&client, &store, cache.as_ref(), status).await
-    }
+        status("Propagating tx...");
+        let propagate_req = RuskRequest::new("propagate_tx", tx_bytes);
+        let _ = self.client.call(2, "Chain", &propagate_req).wait()?;
+        status("Transaction propagated!");
 
-    pub(crate) fn cache(&self) -> Arc<Cache> {
-        let state = self.inner.lock().unwrap();
-        Arc::clone(&state.cache)
+        Ok(utx)
     }
-}
-
-/// Types that are clients of the state API.
-impl StateClient for StateStore {
-    /// Error returned by the node client.
-    type Error = Error;
 
     /// Find notes for a view key, starting from the given block height.
-    fn fetch_notes(
+    pub(crate) fn inputs(
+        &self,
+        index: u8,
+        target: u64,
+    ) -> Result<Vec<(Note, Opening)>, Error> {
+        if let Some((_, vk, pk)) = self.addresses.get(index as usize) {
+            let inputs: Result<Inputs, Error> = self
+                .cache()
+                .notes(pk)?
+                .into_iter()
+                .map(|data| {
+                    let note = data.note;
+                    let value = note.value(Some(vk)).unwrap();
+
+                    Ok((note, value))
+                })
+                .collect();
+
+            try_input_notes(inputs?, target)?
+                .into_iter()
+                .map(|(note, _)| {
+                    let opening = self.fetch_opening(&note)?;
+
+                    Ok((note, opening))
+                })
+                .collect()
+        } else {
+            Err(Error::AddressNotOwned)
+        }
+    }
+
+    pub(crate) fn fetch_notes(
```

Comment:
shouldn't this method fetch the notes for a given `ViewKey` from the network? Here we only get it from the cache.

Reply:
That's the right behaviour.

Indeed in our cache there are only notes that are already been checked used the viewkey.

Basically we "fetch notes from the wallet store for a specific PublicKey)